src/lib/utils/ct_utils.h

```diff
@@ -422,6 +428,87 @@ class Mask final {
       T m_mask;
 };
 
+/**
+* A CT::Option<T> is either a valid T, or not
+*
+* To maintain constant time behavior a value must always be stored.
+* A CT::Choice tracks is the value is valid or not. It is not possible
+* to access the inner value if the Choice is unset.
+*/
+template <typename T>
+class Option {
+   public:
+      /// Construct an Option which contains the specified value, and is set or not
+      constexpr Option(T v, Choice valid) : m_is_some(valid), m_value(std::move(v)) {}
+
+      /// Construct a set option with the provided value
+      constexpr Option(T v) : Option(std::move(v), Choice::yes()) {}
+
+      /// Construct an unset option with a default inner value
+      constexpr Option()
+         requires std::default_initializable<T>
+            : Option(T(), Choice::no()) {}
+
+      /// Return true if this Option contains a value
+      constexpr Choice has_value() const { return m_is_some; }
+
+      /**
+      * Apply a function to the inner value and return a new Option
+      * which contains that value. This is constant time only if @p f is.
+      *
+      * @note The function will always be called, even if the Option is None. It
+      *       must be prepared to handle any possible state of T.
+      */
+      template <std::invocable<const T&> F>
+      constexpr auto transform(F f) const -> Option<std::invoke_result_t<F, const T&>> {
+         return {f(m_value), m_is_some};
+      }
```

Comment:
Mhh, here the return type might need some extra case. Imagine `F` returns a reference for some reason, then you would end up instantiating an `Option<T&>` which most certainly isn't what we want.

Either do

```suggestion
      template <std::invocable<const T&> F>
      constexpr auto transform(F f) const -> Option<std::remove_cvref_t<std::invoke_result_t<F, const T&>>> {
         return {f(m_value), m_is_some};
      }
```

and/or exclude references for `typename T` in general with

```C++
template <typename T>
   requires !std::is_reference_t<T>
class Option {
```

Reply:
`Option<&T>` might make sense in some scenario. I can't think of an immediate use case though

Reply:
Perhaps. But this is what `std::reference_wrapper<>` is for, IMHO. Otherwise, imagine this:

```C++
auto foo(Option<MyConstTimeType> o) {
   return o.transform([](const MyConstTimeType& ov) {
      return ov.get_some_member_by_reference();
   });
}
```

... `foo()` would return an `Option<T&>` that would most-certainly be a dangling reference, given that the getter returns a reference to a member of the local copy of `Option<T> o`. That's a source of very subtle memory bugs.